src/org/gk/pro/ProExporter.java

```diff
@@ -0,0 +1,643 @@
+
+        return output;
+    }
+
+    /**
+     * Return the modifications for a given IntraChainCrosslinkedResidue.
+     *
+     * Special case of {@link #getModifications(GKInstance, List)}
+     *
+     * @param modifiedResidue
+     * @return String
+     * @throws Exception
+     */
+    private String getIntraChainModifications(List<Object> modifiedResidues) throws Exception {
+        String output = "";
+        String modification = null;
+        IntraChainCrosslinkedResidue residueExporter = new IntraChainCrosslinkedResidue();
+        for (Object modifiedResidue : modifiedResidues) {
+            modification = residueExporter.exportPsiModIdentifier((GKInstance) modifiedResidue);
+            if (modification != null)
+                output += modification;
+        }
+        resetIndices();
+
+        for (Object modifiedResidue : modifiedResidues) {
+            modification = residueExporter.exportModificationIdentifier((GKInstance) modifiedResidue);
+            if (modification != null)
+                output += modification;
+        }
+        resetIndices();
+
+        return output;
+    }
+
+    /**
+     * Return the freeText values for a given EWAS.
+     *
+     * @param modifiedResidues
+     * @return String
+     * @throws InvalidAttributeException
+     * @throws Exception
+     */
+    private String getFreeText(List<Object> modifiedResidues) throws InvalidAttributeException, Exception {
+        if (modifiedResidues == null || modifiedResidues.size() == 0)
+            return null;
+        String output = "";
+        AbstractModifiedResidue residueExporter = null;
+        String pkg = this.getClass().getPackage().getName();
+        String cls = null;
+        // Modification classes that have free text columns.
+        List<String> allowedClasses = Arrays.asList(ReactomeJavaConstants.FragmentInsertionModification,
+                                                    ReactomeJavaConstants.FragmentDeletionModification,
+                                                    ReactomeJavaConstants.FragmentReplacedModification,
+                                                    ReactomeJavaConstants.IntraChainCrosslinkedResidue);
+        for (Object modifiedResidue : modifiedResidues) {
+            if (allowedClasses.stream().noneMatch(((GKInstance) modifiedResidue).getSchemClass()::isa))
+                continue;
+
+            if (output.length() > 0)
+                output += ProExporterConstants.freeTextDelimiter;
+
+            cls = ((GKInstance) modifiedResidue).getSchemClass().getName();
+            residueExporter = (AbstractModifiedResidue) Class.forName(pkg + "." + cls).getConstructor().newInstance();
+            output += residueExporter.exportFreeText((GKInstance) modifiedResidue);
+        }
+        resetIndices();
+
+        return output;
+    }
+
+    /**
+     * Helper function to reset indices for FragmentModification residues.
+     */
+    private void resetIndices() {
```

Comment:
Are you sure you have to use static methods for reset indices? In the object scope methods, using static methods is just not good programming.

Reply:
That was not a great decision on my part -- the index "logic" is now handled directly in the ProExporter class.